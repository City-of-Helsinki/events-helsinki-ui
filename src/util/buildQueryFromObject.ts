/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
import forEach from 'lodash/forEach';
import isArray from 'lodash/isArray';
import isEmpty from 'lodash/isEmpty';
import isNil from 'lodash/isNil';
import isNumber from 'lodash/isNumber';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const buildQueryFromObject = (obj: { [key: string]: any }) => {
  const query: string[] = [];

  forEach(obj, (filter, key) => {
    if (!isEmpty(filter) || isNumber(filter) || typeof filter === 'boolean') {
      if (isArray(filter)) {
        const items: Array<string | number> = [];

        forEach(filter, (item: string | number) => {
          items.push(encodeURIComponent(item));
        });

        query.push(`${key}=${items.join(',')}`);
      } else if (!isNil(filter)) {
        query.push(`${key}=${encodeURIComponent(filter)}`);
      }
    }
  });

  return query.length ? `?${query.join('&')}` : '';
};

export default buildQueryFromObject;
