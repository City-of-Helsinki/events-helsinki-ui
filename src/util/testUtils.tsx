import { fireEvent } from '@testing-library/react';

export const arrowUpKeyPressHelper = () =>
  fireEvent.keyDown(document, { code: 38, key: 'ArrowUp' });

export const arrowDownKeyPressHelper = () =>
  fireEvent.keyDown(document, { code: 40, key: 'ArrowDown' });

export const escKeyPressHelper = () =>
  fireEvent.keyDown(document, { code: 27, key: 'Escape' });

export const tabKeyPressHelper = () =>
  fireEvent.keyDown(document, { code: 9, key: 'Tab' });
