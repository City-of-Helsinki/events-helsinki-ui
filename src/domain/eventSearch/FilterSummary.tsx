import get from "lodash/get";
import React from "react";
import { useTranslation } from "react-i18next";
import { useHistory, useLocation } from "react-router";

import FilterButton, {
  FilterType
} from "../../common/components/filterButton/FilterButton";
import { DISTRICTS, TARGET_GROUPS } from "../../constants";
import useLocale from "../../hooks/useLocale";
import { formatDate } from "../../util/dateUtils";
import getUrlParamAsString from "../../util/getUrlParamAsString";
import { getSearchQuery } from "../../util/searchUtils";
import { translateValue } from "../../util/translateUtils";
import DateFilter from "./DateFilter";
import styles from "./filterSummary.module.scss";
import KeywordFilter from "./KeywordFilter";
import PlaceFilter from "./PlaceFilter";
import PublisherFilter from "./PublisherFilter";

const findKeyOfDistrict = (value: string) =>
  Object.keys(DISTRICTS).find(key => get(DISTRICTS, key) === value);

const findKeyOfTarget = (value: string) =>
  Object.keys(TARGET_GROUPS).find(key => get(TARGET_GROUPS, key) === value);

const FilterSummary = () => {
  const { t } = useTranslation();
  const locale = useLocale();
  const { push } = useHistory();
  const searchParams = new URLSearchParams(useLocation().search);
  const publisher = searchParams.get("publisher");
  const categories = getUrlParamAsString(searchParams, "categories");
  const dateTypes = getUrlParamAsString(searchParams, "dateTypes");
  const districts = getUrlParamAsString(searchParams, "districts");
  const keywords = getUrlParamAsString(searchParams, "keywords");
  const places = getUrlParamAsString(searchParams, "places");
  const targets = getUrlParamAsString(searchParams, "targets");
  const startDate = searchParams.get("startDate");
  const endDate = searchParams.get("endDate");
  const dateText = startDate
    ? endDate
      ? `${formatDate(new Date(startDate))} - ${formatDate(new Date(endDate))}`
      : formatDate(new Date(startDate))
    : endDate
    ? formatDate(new Date(endDate))
    : "";

  const handleFilterRemove = (value: string, type: FilterType) => {
    const endDate = searchParams.get("endDate");
    const startDate = searchParams.get("startDate");

    const search = getSearchQuery({
      categories:
        type === "category"
          ? categories.filter(category => category !== value)
          : categories,
      dateTypes:
        type === "dateType"
          ? dateTypes.filter(dateType => dateType !== value)
          : dateTypes,
      districts:
        type === "district"
          ? districts.filter(district => district !== value)
          : districts,
      endDate: type === "date" ? null : endDate ? new Date(endDate) : null,
      isCustomDate: !!(startDate || endDate),
      keywords:
        type === "keyword" || type === "yso"
          ? keywords.filter(keyword => keyword !== value)
          : keywords,
      places:
        type === "place" ? places.filter(place => place !== value) : places,
      publisher: type !== "publisher" ? searchParams.get("publisher") : null,
      search: "",
      startDate:
        type === "date" ? null : startDate ? new Date(startDate) : null,
      targets:
        type === "target" ? targets.filter(target => target !== value) : targets
    });

    push({ pathname: `/${locale}/events`, search });
  };

  const clearFilters = () => {
    const search = getSearchQuery({
      categories: [],
      dateTypes: [],
      districts: [],
      endDate: null,
      isCustomDate: !!(startDate || endDate),
      keywords: [],
      places: [],
      publisher: null,
      search: "",
      startDate: null,
      targets: []
    });

    push({ pathname: `/${locale}/events`, search });
  };

  const hasFilters =
    !!searchParams.get("publisher") ||
    !!categories.length ||
    !!dateText ||
    !!dateTypes.length ||
    !!districts.length ||
    !!keywords.length ||
    !!places.length ||
    !!targets.length;

  return (
    <div className={styles.filterSummary}>
      <div className={styles.filtersWrapper}>
        <h4 className={styles.titleFilterSummary}>
          {t("eventSearch.filters.titleSummary")}
        </h4>
        {/* TODO: Add filters summary here */}
        {hasFilters ? (
          <>
            {categories.map(category => (
              <FilterButton
                key={category}
                onRemove={handleFilterRemove}
                text={translateValue("home.category.", category, t)}
                type="category"
                value={category}
              />
            ))}
            {keywords.map(keyword => (
              <KeywordFilter
                key={keyword}
                onRemove={handleFilterRemove}
                id={keyword}
              />
            ))}
            {publisher && (
              <PublisherFilter id={publisher} onRemove={handleFilterRemove} />
            )}
            {dateText && (
              <DateFilter
                onRemove={handleFilterRemove}
                text={dateText}
                type="date"
                value="date"
              />
            )}
            {dateTypes.map(dateType => (
              <DateFilter
                key={dateType}
                onRemove={handleFilterRemove}
                type="dateType"
                value={dateType}
              />
            ))}
            {districts.map(district => (
              <FilterButton
                key={district}
                onRemove={handleFilterRemove}
                text={translateValue(
                  "commons.districts.",
                  findKeyOfDistrict(district) || "",
                  t
                )}
                type="district"
                value={district}
              />
            ))}
            {places.map(place => (
              <PlaceFilter
                key={place}
                id={place}
                onRemove={handleFilterRemove}
              />
            ))}
            {targets.map(target => (
              <FilterButton
                key={target}
                onRemove={handleFilterRemove}
                text={translateValue(
                  "commons.targets.",
                  findKeyOfTarget(target) || "",
                  t
                )}
                type="target"
                value={target}
              />
            ))}
            <button
              className={styles.clearButton}
              onClick={clearFilters}
              type="button"
            >
              {t("eventSearch.buttonClearFilters")}
            </button>
          </>
        ) : (
          t("eventSearch.filters.textNoFilters")
        )}
      </div>
    </div>
  );
};

export default FilterSummary;
