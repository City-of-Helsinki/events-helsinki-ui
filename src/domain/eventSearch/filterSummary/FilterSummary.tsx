import React from 'react';
import { useTranslation } from 'react-i18next';
import { useHistory, useLocation } from 'react-router';

import FilterButton, {
  FilterType,
} from '../../../common/components/filterButton/FilterButton';
import useDivisionOptions from '../../../hooks/useDivisionOptions';
import useLocale from '../../../hooks/useLocale';
import { formatDate } from '../../../util/dateUtils';
import { translateValue } from '../../../util/translateUtils';
import { ROUTES } from '../../app/routes/constants';
import {
  getSearchFilters,
  getSearchQuery,
  removeSuitableForFilterValue,
} from '../utils';
import AgeFilter from './AgeFilter';
import DateFilter from './DateFilter';
import styles from './filterSummary.module.scss';
import PlaceFilter from './PlaceFilter';
import PublisherFilter from './PublisherFilter';
import TextFilter from './TextFilter';

export const filterSummaryContainerTestId = 'filter-summary';

interface Props {
  onClear: () => void;
  route?: '/events' | '/courses';
}

const FilterSummary: React.FC<Props> = ({ onClear, route }) => {
  const { t } = useTranslation();
  const locale = useLocale();
  const { push } = useHistory();
  const searchParams = new URLSearchParams(useLocation().search);
  const {
    categories,
    dateTypes,
    divisions,
    end,
    isFree,
    keyword,
    keywordNot,
    onlyChildrenEvents,
    places,
    publisher,
    start,
    text,
    suitableFor,
    hobbyTypes,
  } = getSearchFilters(searchParams);

  const [minAge, maxAge] = suitableFor ?? [];

  const dateText =
    start || end
      ? `${start ? formatDate(start) : ''} - ${
          end ? formatDate(end) : ''
        }`.trim()
      : '';

  const neighborhoods = useDivisionOptions();
  const getNeighorhoodName = React.useCallback(
    (id: string) => {
      const neighborhood = neighborhoods.find((item) => item.value === id);
      return neighborhood?.text ?? '';
    },
    [neighborhoods]
  );

  const handleFilterRemove = (value: string | number, type: FilterType) => {
    const getFilteredList = (listType: FilterType, list: string[] = []) =>
      type === listType ? list.filter((v) => v !== value) : list;

    const search = getSearchQuery({
      categories: getFilteredList('category', categories),
      hobbyTypes: getFilteredList('hobbyType', hobbyTypes),
      dateTypes: getFilteredList('dateType', dateTypes),
      divisions: getFilteredList('division', divisions),
      end: type === 'date' ? null : end,
      isFree,
      keyword,
      keywordNot,
      onlyChildrenEvents,
      places: getFilteredList('place', places),
      publisher: type !== 'publisher' ? publisher : null,
      start: type === 'date' ? null : start,
      text: getFilteredList('text', text),
      suitableFor: removeSuitableForFilterValue(suitableFor, type),
    });

    push({ pathname: `/${locale}${route || ROUTES.EVENTS}`, search });
  };

  const hasFilters =
    !!publisher ||
    !!categories.length ||
    !!hobbyTypes?.length ||
    !!dateText ||
    !!dateTypes.length ||
    !!divisions.length ||
    !!places.length ||
    !!text.length ||
    !!suitableFor?.length;

  if (!hasFilters) return null;

  return (
    <div
      className={styles.filterSummary}
      data-testid={filterSummaryContainerTestId}
    >
      {text.map((item, index) => (
        <TextFilter key={index} text={item} onRemove={handleFilterRemove} />
      ))}
      {categories.map((category) => (
        <FilterButton
          key={category}
          onRemove={handleFilterRemove}
          text={translateValue(
            `home.category.${route === '/courses' ? 'courses.' : ''}`,
            category,
            t
          )}
          type="category"
          value={category}
        />
      ))}
      {hobbyTypes?.map((hobbyType) => (
        <FilterButton
          key={hobbyType}
          onRemove={handleFilterRemove}
          text={translateValue('home.hobby.', hobbyType, t)}
          type="hobbyType"
          value={hobbyType}
        />
      ))}
      {publisher && (
        <PublisherFilter id={publisher} onRemove={handleFilterRemove} />
      )}
      {divisions.map((division) => (
        <FilterButton
          key={division}
          onRemove={handleFilterRemove}
          text={getNeighorhoodName(division)}
          type="division"
          value={division}
        />
      ))}
      {places.map((place) => (
        <PlaceFilter key={place} id={place} onRemove={handleFilterRemove} />
      ))}

      {dateText && (
        <DateFilter
          onRemove={handleFilterRemove}
          text={dateText}
          type="date"
          value="date"
        />
      )}
      {dateTypes.map((dateType) => (
        <DateFilter
          key={dateType}
          onRemove={handleFilterRemove}
          type="dateType"
          value={dateType}
        />
      ))}
      {minAge != null && (
        <AgeFilter
          type="minAge"
          value={minAge?.toString()}
          onRemove={handleFilterRemove}
        />
      )}
      {maxAge != null && (
        <AgeFilter
          type="maxAge"
          value={maxAge?.toString()}
          onRemove={handleFilterRemove}
        />
      )}
      <button className={styles.clearButton} onClick={onClear} type="button">
        {t('eventSearch.buttonClearFilters')}
      </button>
    </div>
  );
};

export default FilterSummary;
