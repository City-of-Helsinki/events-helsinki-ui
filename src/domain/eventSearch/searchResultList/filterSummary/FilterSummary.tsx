import get from "lodash/get";
import React from "react";
import { useTranslation } from "react-i18next";
import { useHistory, useLocation } from "react-router";

import FilterButton, {
  FilterType
} from "../../../../common/components/filterButton/FilterButton";
import { TARGET_GROUPS } from "../../../../constants";
import { useNeighborhoodListQuery } from "../../../../generated/graphql";
import useLocale from "../../../../hooks/useLocale";
import { formatDate } from "../../../../util/dateUtils";
import getLocalisedString from "../../../../util/getLocalisedString";
import getUrlParamAsString from "../../../../util/getUrlParamAsString";
import { getSearchQuery } from "../../../../util/searchUtils";
import { translateValue } from "../../../../util/translateUtils";
import DateFilter from "./DateFilter";
import styles from "./filterSummary.module.scss";
import KeywordFilter from "./KeywordFilter";
import PlaceFilter from "./PlaceFilter";
import PublisherFilter from "./PublisherFilter";
import SearchWordFilter from "./SearchWordFilter";

const findKeyOfTarget = (value: string) =>
  Object.keys(TARGET_GROUPS).find(key => get(TARGET_GROUPS, key) === value);

const FilterSummary = () => {
  const { t } = useTranslation();
  const locale = useLocale();
  const { push } = useHistory();
  const searchParams = new URLSearchParams(useLocation().search);
  const publisher = searchParams.get("publisher");
  const categories = getUrlParamAsString(searchParams, "categories");
  const dateTypes = getUrlParamAsString(searchParams, "dateTypes");
  const districts = getUrlParamAsString(searchParams, "districts");
  const isFree = searchParams.get("isFree") === "true" ? true : false;
  const keywords = getUrlParamAsString(searchParams, "keywords");
  const places = getUrlParamAsString(searchParams, "places");
  const targets = getUrlParamAsString(searchParams, "targets");
  const startDate = searchParams.get("startDate");
  const endDate = searchParams.get("endDate");
  const searchWord = searchParams.get("search");
  const dateText =
    startDate || endDate
      ? `${startDate ? formatDate(new Date(startDate)) : ""} - ${
          endDate ? formatDate(new Date(endDate)) : ""
        }`.trim()
      : "";

  const { data: neighborhoodsData } = useNeighborhoodListQuery();

  const getNeighorhoodName = React.useCallback(
    (id: string) => {
      const neighborhoods = neighborhoodsData
        ? neighborhoodsData.neighborhoodList.data
        : [];
      const neighborhood = neighborhoods.find(item => item.id === id);
      return getLocalisedString(neighborhood ? neighborhood.name : {}, locale);
    },
    [locale, neighborhoodsData]
  );

  const handleFilterRemove = (value: string, type: FilterType) => {
    const endDate = searchParams.get("endDate");
    const searchWord = searchParams.get("search") || "";
    const startDate = searchParams.get("startDate");

    const search = getSearchQuery({
      categories:
        type === "category"
          ? categories.filter(category => category !== value)
          : categories,
      dateTypes:
        type === "dateType"
          ? dateTypes.filter(dateType => dateType !== value)
          : dateTypes,
      districts:
        type === "district"
          ? districts.filter(district => district !== value)
          : districts,
      endDate: type === "date" ? null : endDate ? new Date(endDate) : null,
      isFree,
      keywords:
        type === "keyword" || type === "yso"
          ? keywords.filter(keyword => keyword !== value)
          : keywords,
      places:
        type === "place" ? places.filter(place => place !== value) : places,
      publisher: type !== "publisher" ? searchParams.get("publisher") : null,
      search: type === "searchWord" ? "" : searchWord,
      startDate:
        type === "date" ? null : startDate ? new Date(startDate) : null,
      targets:
        type === "target" ? targets.filter(target => target !== value) : targets
    });

    push({ pathname: `/${locale}/events`, search });
  };

  const deleteSearchWord = () => {
    handleFilterRemove("", "searchWord");
  };

  const clearFilters = () => {
    const search = getSearchQuery({
      categories: [],
      dateTypes: [],
      districts: [],
      endDate: null,
      isFree: false,
      keywords: [],
      places: [],
      publisher: null,
      search: "",
      startDate: null,
      targets: []
    });

    push({ pathname: `/${locale}/events`, search });
  };

  const hasFilters =
    !!searchParams.get("publisher") ||
    !!categories.length ||
    !!dateText ||
    !!dateTypes.length ||
    !!districts.length ||
    !!keywords.length ||
    !!places.length ||
    !!searchWord ||
    !!targets.length;

  return (
    <div className={styles.filterSummary}>
      <div className={styles.filtersWrapper}>
        <h2 className={styles.titleFilterSummary}>
          {t("eventSearch.filters.titleSummary")}
        </h2>
        {!!searchWord && (
          <SearchWordFilter
            onRemove={deleteSearchWord}
            searchWord={searchWord}
          />
        )}
        {hasFilters ? (
          <>
            {categories.map(category => (
              <FilterButton
                key={category}
                onRemove={handleFilterRemove}
                text={translateValue("home.category.", category, t)}
                type="category"
                value={category}
              />
            ))}
            {keywords.map(keyword => (
              <KeywordFilter
                key={keyword}
                onRemove={handleFilterRemove}
                id={keyword}
              />
            ))}
            {publisher && (
              <PublisherFilter id={publisher} onRemove={handleFilterRemove} />
            )}
            {districts.map(district => (
              <FilterButton
                key={district}
                onRemove={handleFilterRemove}
                text={getNeighorhoodName(district)}
                type="district"
                value={district}
              />
            ))}
            {places.map(place => (
              <PlaceFilter
                key={place}
                id={place}
                onRemove={handleFilterRemove}
              />
            ))}
            {targets.map(target => (
              <FilterButton
                key={target}
                onRemove={handleFilterRemove}
                text={translateValue(
                  "commons.targets.",
                  findKeyOfTarget(target) || "",
                  t
                )}
                type="target"
                value={target}
              />
            ))}
            {dateText && (
              <DateFilter
                onRemove={handleFilterRemove}
                text={dateText}
                type="date"
                value="date"
              />
            )}
            {dateTypes.map(dateType => (
              <DateFilter
                key={dateType}
                onRemove={handleFilterRemove}
                type="dateType"
                value={dateType}
              />
            ))}
            <button
              className={styles.clearButton}
              onClick={clearFilters}
              type="button"
            >
              {t("eventSearch.buttonClearFilters")}
            </button>
          </>
        ) : (
          t("eventSearch.filters.textNoFilters")
        )}
      </div>
    </div>
  );
};

export default FilterSummary;
