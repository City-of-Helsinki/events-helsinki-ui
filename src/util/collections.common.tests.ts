import { MockedResponse } from '@apollo/react-testing';
import { screen, waitFor } from '@testing-library/react';
import range from 'lodash/range';

import {
  CollectionListDocument,
  CollectionListQueryVariables,
  CollectionListResponse,
} from '../generated/graphql';
import { fakeCollections } from './mockDataUtils';
import { render } from './testUtils';

export const getCollectionQueryListMocks = (
  collections: CollectionListResponse,
  variables?: CollectionListQueryVariables
): MockedResponse[] => [
  {
    request: {
      query: CollectionListDocument,
      variables,
    },
    result: { data: { collectionList: collections } },
  },
];

const renderAndWaitForComponentToBeLoaded = async ({
  component,
  collections,
  mocks = [],
  variables,
}: {
  component: React.ReactElement;
  collections: CollectionListResponse;
  mocks: MockedResponse[];
  variables?: CollectionListQueryVariables;
}) => {
  const collectionMocks = getCollectionQueryListMocks(collections, variables);
  render(component, { mocks: [...collectionMocks, ...mocks] });

  await waitFor(() => {
    expect(screen.queryByTestId('loading-spinner')).not.toBeInTheDocument();
  });
};

const renderComponentAndExpectOddCollectionsToBePresent = async ({
  component,
  collections,
  collectionNames,
  mocks = [],
  variables,
}: {
  component: React.ReactElement;
  collections: CollectionListResponse;
  collectionNames: string[];
  mocks?: MockedResponse[];
  variables?: CollectionListQueryVariables;
}) => {
  await renderAndWaitForComponentToBeLoaded({
    component,
    collections,
    mocks,
    variables,
  });

  collections.data.forEach((collection, index) => {
    if (index % 2) {
      expect(
        screen.queryByText(collectionNames[index])
      ).not.toBeInTheDocument();
    } else {
      expect(screen.getByText(collectionNames[index])).toBeInTheDocument();
    }
  });
};

export const collectionListFilterTests = ({
  component,
  generatedCollectionListSize = 7,
  mocks = [],
  variables,
}: {
  component: React.ReactElement;
  generatedCollectionListSize?: number;
  mocks?: MockedResponse[];
  variables?: CollectionListQueryVariables;
}): void => {
  const collectionIds = range(generatedCollectionListSize).map((id) =>
    (id + 1).toString()
  );
  const collectionNames = collectionIds.map((id) => `Collection ${id}`);

  it('should show all collections', async () => {
    const collections = fakeCollections(
      collectionIds.length,
      collectionIds.map((id, index) => ({
        id,
        title: { fi: collectionNames[index] },
      }))
    );
    await renderAndWaitForComponentToBeLoaded({
      component,
      collections,
      mocks,
      variables,
    });

    collections.data.forEach((collection) => {
      expect(screen.getByText(collection.title.fi)).toBeInTheDocument();
    });
  });

  it('should not show collection if language is not supported', async () => {
    const collections = fakeCollections(
      collectionIds.length,
      collectionIds.map((id, index) => ({
        id,
        title: { fi: Boolean(1 - (index % 2)) ? collectionNames[index] : '' },
      }))
    );

    await renderComponentAndExpectOddCollectionsToBePresent({
      component,
      collections,
      collectionNames,
      mocks,
      variables,
    });
  });

  it('should not show expired collections', async () => {
    const collections = fakeCollections(
      collectionIds.length,
      collectionIds.map((id, index) => ({
        id,
        expired: Boolean(index % 2),
        title: { fi: collectionNames[index] },
      }))
    );

    await renderComponentAndExpectOddCollectionsToBePresent({
      component,
      collections,
      collectionNames,
      mocks,
      variables,
    });
  });

  it('should not show collections that are not live', async () => {
    const collections = fakeCollections(
      collectionIds.length,
      collectionIds.map((id, index) => ({
        id,
        live: Boolean(1 - (index % 2)),
        title: { fi: collectionNames[index] },
      }))
    );
    await renderComponentAndExpectOddCollectionsToBePresent({
      component,
      collections,
      collectionNames,
      mocks,
      variables,
    });
  });
};
