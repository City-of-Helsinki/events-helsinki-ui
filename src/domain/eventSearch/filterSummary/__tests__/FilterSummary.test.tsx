import { screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';

import translations from '../../../../common/translation/i18n/fi.json';
import {
  NeighborhoodListDocument,
  OrganizationDetailsDocument,
  PlaceDetailsDocument,
} from '../../../../generated/graphql';
import { render } from '../../../../util/testUtils';
import neighborhoodListResponse from '../../../neighborhood/__mocks__/neighborhoodListResponse';
import mockOrganization from '../../../organisation/__mocks__/organizationDetails';
import mockPlace from '../../../place/__mocks__/place';
import FilterSummary from '../FilterSummary';

const mocks = [
  {
    request: {
      query: NeighborhoodListDocument,
    },
    result: neighborhoodListResponse,
  },
  {
    request: {
      query: OrganizationDetailsDocument,
      variables: {
        id: mockOrganization.id,
      },
    },
    result: {
      data: {
        organizationDetails: mockOrganization,
      },
    },
  },
  {
    request: {
      query: PlaceDetailsDocument,
      variables: {
        id: mockPlace.id,
      },
    },
    result: {
      data: {
        placeDetails: mockPlace,
      },
    },
  },
];

interface UrlParams {
  categories: string;
  dateTypes: string;
  divisions: string;
  end: string;
  places: string;
  publisher: string;
  start: string;
  text: string;
}

const urlParams: UrlParams = {
  categories: 'movie',
  dateTypes: 'today',
  divisions: 'kaupunginosa%3Aalppiharju',
  end: '2020-08-23',
  places: mockPlace.id as string,
  publisher: mockOrganization.id as string,
  start: '2020-08-20',
  text: 'jazz',
};

type UrlParamKeys = keyof UrlParams;

const routes = [
  // eslint-disable-next-line max-len
  `/fi/events?categories=${urlParams.categories}&dateTypes=today&divisions=${urlParams.divisions}&end=${urlParams.end}&places=${urlParams.places}&publisher=${urlParams.publisher}&start=${urlParams.start}&text=${urlParams.text}`,
];

it('matches snapshot', async () => {
  const { container } = render(<FilterSummary onClear={jest.fn()} />, {
    mocks,
    routes,
  });

  await waitFor(() => {
    expect(
      screen.queryByText((mockPlace.name || {})['fi'] || '')
    ).toBeInTheDocument();
  });

  expect(container.firstChild).toMatchSnapshot();
});

it('calls onClear callback when clear button is clicked ', async () => {
  const onClear = jest.fn();
  render(<FilterSummary onClear={onClear} />, {
    mocks,
    routes,
  });

  await waitFor(() => {
    expect(
      screen.queryByText((mockPlace.name || {})['fi'] || '')
    ).toBeDefined();
  });

  const button = screen.getByRole('button', {
    name: translations.eventSearch.buttonClearFilters,
  });

  userEvent.click(button);
  expect(onClear).toBeCalledTimes(1);
});

it('routes to correct url after deleting filters ', async () => {
  const { history } = render(<FilterSummary onClear={jest.fn()} />, {
    mocks,
    routes,
  });

  await waitFor(() => {
    expect(
      screen.queryByText((mockPlace.name || {})['fi'] || '')
    ).toBeDefined();
  });

  const items: { button: string; params: UrlParamKeys[] }[] = [
    { button: 'Poista suodatin: T채n채채n', params: ['dateTypes'] },
    {
      button: 'Poista suodatin: 20.08.2020 - 23.08.2020',
      params: ['end', 'start'],
    },
    { button: 'Poista suodatin: Elokuva', params: ['categories'] },
    { button: 'Poista suodatin: Alppiharju', params: ['divisions'] },
    { button: 'Poista suodatin: Gr채san taitojen talo', params: ['places'] },
    {
      button: 'Poista suodatin: Yleiset kulttuuripalvelut',
      params: ['publisher'],
    },
    { button: 'Poista suodatin: jazz', params: ['text'] },
  ];

  items.forEach(item => {
    item.params.forEach(param => {
      expect(new URLSearchParams(history.location.search).get(param)).toBe(
        decodeURIComponent(urlParams[param] || '')
      );
    });

    userEvent.click(
      screen.getByRole('button', {
        name: item.button,
      })
    );

    item.params.forEach(param => {
      expect(
        new URLSearchParams(history.location.search).get(param)
      ).toBeNull();
    });
  });
});
