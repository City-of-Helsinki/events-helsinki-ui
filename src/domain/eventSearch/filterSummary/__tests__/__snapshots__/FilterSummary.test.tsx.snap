// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  class="filterSummary"
>
  <div
    class="filter text"
  >
    <button
      aria-label="Poista suodatin: jazz"
      class="closeButton"
      type="button"
    >
      <svg
        class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_s__2WGWe icon_hds-icon--size-s__2Lkik"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          fill="none"
          fill-rule="evenodd"
        >
          <path
            d="M0 0h24v24H0z"
          />
          <path
            d="M17.657 7.757L13.414 12l4.243 4.243-1.414 1.414L12 13.414l-4.243 4.243-1.414-1.414L10.586 12 6.343 7.757l1.414-1.414L12 10.586l4.243-4.243z"
            fill="currentColor"
          />
        </g>
      </svg>
    </button>
    jazz
  </div>
  <div
    class="filter category"
  >
    <button
      aria-label="Poista suodatin: Elokuva"
      class="closeButton"
      type="button"
    >
      <svg
        class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_s__2WGWe icon_hds-icon--size-s__2Lkik"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          fill="none"
          fill-rule="evenodd"
        >
          <path
            d="M0 0h24v24H0z"
          />
          <path
            d="M17.657 7.757L13.414 12l4.243 4.243-1.414 1.414L12 13.414l-4.243 4.243-1.414-1.414L10.586 12 6.343 7.757l1.414-1.414L12 10.586l4.243-4.243z"
            fill="currentColor"
          />
        </g>
      </svg>
    </button>
    Elokuva
  </div>
  <div
    class="filter publisher"
  >
    <button
      aria-label="Poista suodatin: Yleiset kulttuuripalvelut"
      class="closeButton"
      type="button"
    >
      <svg
        class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_s__2WGWe icon_hds-icon--size-s__2Lkik"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          fill="none"
          fill-rule="evenodd"
        >
          <path
            d="M0 0h24v24H0z"
          />
          <path
            d="M17.657 7.757L13.414 12l4.243 4.243-1.414 1.414L12 13.414l-4.243 4.243-1.414-1.414L10.586 12 6.343 7.757l1.414-1.414L12 10.586l4.243-4.243z"
            fill="currentColor"
          />
        </g>
      </svg>
    </button>
    Yleiset kulttuuripalvelut
  </div>
  <div
    class="filter division"
  >
    <button
      aria-label="Poista suodatin: Alppiharju"
      class="closeButton"
      type="button"
    >
      <svg
        class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_s__2WGWe icon_hds-icon--size-s__2Lkik"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          fill="none"
          fill-rule="evenodd"
        >
          <path
            d="M0 0h24v24H0z"
          />
          <path
            d="M17.657 7.757L13.414 12l4.243 4.243-1.414 1.414L12 13.414l-4.243 4.243-1.414-1.414L10.586 12 6.343 7.757l1.414-1.414L12 10.586l4.243-4.243z"
            fill="currentColor"
          />
        </g>
      </svg>
    </button>
    Alppiharju
  </div>
  <div
    class="filter place"
  >
    <button
      aria-label="Poista suodatin: Gräsan taitojen talo"
      class="closeButton"
      type="button"
    >
      <svg
        class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_s__2WGWe icon_hds-icon--size-s__2Lkik"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          fill="none"
          fill-rule="evenodd"
        >
          <path
            d="M0 0h24v24H0z"
          />
          <path
            d="M17.657 7.757L13.414 12l4.243 4.243-1.414 1.414L12 13.414l-4.243 4.243-1.414-1.414L10.586 12 6.343 7.757l1.414-1.414L12 10.586l4.243-4.243z"
            fill="currentColor"
          />
        </g>
      </svg>
    </button>
    Gräsan taitojen talo
  </div>
  <div
    class="filter date"
  >
    <button
      aria-label="Poista suodatin: 20.08.2020 - 23.08.2020"
      class="closeButton"
      type="button"
    >
      <svg
        class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_s__2WGWe icon_hds-icon--size-s__2Lkik"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          fill="none"
          fill-rule="evenodd"
        >
          <path
            d="M0 0h24v24H0z"
          />
          <path
            d="M17.657 7.757L13.414 12l4.243 4.243-1.414 1.414L12 13.414l-4.243 4.243-1.414-1.414L10.586 12 6.343 7.757l1.414-1.414L12 10.586l4.243-4.243z"
            fill="currentColor"
          />
        </g>
      </svg>
    </button>
    20.08.2020 - 23.08.2020
  </div>
  <div
    class="filter dateType"
  >
    <button
      aria-label="Poista suodatin: Tänään"
      class="closeButton"
      type="button"
    >
      <svg
        class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_s__2WGWe icon_hds-icon--size-s__2Lkik"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          fill="none"
          fill-rule="evenodd"
        >
          <path
            d="M0 0h24v24H0z"
          />
          <path
            d="M17.657 7.757L13.414 12l4.243 4.243-1.414 1.414L12 13.414l-4.243 4.243-1.414-1.414L10.586 12 6.343 7.757l1.414-1.414L12 10.586l4.243-4.243z"
            fill="currentColor"
          />
        </g>
      </svg>
    </button>
    Tänään
  </div>
  <button
    class="clearButton"
    type="button"
  >
    Tyhjennä hakuehdot
  </button>
</div>
`;
